### PR TITLE
Plugin homebrew

### PR DESCRIPTION
**Pull request by jcftang**
This a first pass at teaching bcfg2 about homebrew on OSX. Much of this is lifted from the MacPorts.py file.

### CODE REVIEW

**Comment by AlexanderS** [comment]
You may want to drop the first 6 commits from this pull request, because they are already in master.

File: src/lib/Bcfg2/Client/Tools/HomeBrew.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+    def RefreshPackages(self):
+        """Refresh memory hashes of packages."""
+        pkgcache = self.cmd.run(["/usr/local/bin/brew",
+                                 "list", "--versions"]).stdout.splitlines()
+        self.installed = {}
+        for pkg in pkgcache:
+            pkgname = pkg.split(' ')[0].strip()
+            version = pkg.split(' ')[1].strip()
+            self.logger.info(" pkgname: %s version: %s" % (pkgname, version))
```
Comment (stpierre) [inline src/lib/Bcfg2/Client/Tools/HomeBrew.py]: This looks like it should be logged at a debug level, if at all.

File: src/lib/Bcfg2/Client/Tools/HomeBrew.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        self.RefreshPackages()
+
+    def RefreshPackages(self):
+        """Refresh memory hashes of packages."""
+        pkgcache = self.cmd.run(["/usr/local/bin/brew",
+                                 "list", "--versions"]).stdout.splitlines()
+        self.installed = {}
+        for pkg in pkgcache:
+            pkgname = pkg.split(' ')[0].strip()
+            version = pkg.split(' ')[1].strip()
```
Comment (stpierre) [inline src/lib/Bcfg2/Client/Tools/HomeBrew.py]: ``pkgname, version = pkg.strip().split()``

Or, if version might contain spaces:

``pkgname, version = pkg.strip().split(None, 2)``

File: src/lib/Bcfg2/Client/Tools/HomeBrew.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            self.logger.info(" pkgname: %s version: %s" % (pkgname, version))
+            self.installed[pkgname] = version
+
+    def VerifyPackage(self, entry, _):
+        """Verify Package status for entry."""
+	self.logger.info("VerifyPackage: %s : %s" % (entry.get('name'),
+                                                     entry.get('version')))
```
Comment (stpierre) [inline src/lib/Bcfg2/Client/Tools/HomeBrew.py]: Log this at debug.

File: src/lib/Bcfg2/Client/Tools/HomeBrew.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        if 'version' not in entry.attrib:
```
Comment (stpierre) [inline src/lib/Bcfg2/Client/Tools/HomeBrew.py]: This block of code is not reachable; ``Verify*`` is only called if ``canVerify`` succeeds, and ``canVerify`` checks that all attributes in ``self.__req__`` are present.

File: src/lib/Bcfg2/Client/Tools/HomeBrew.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+                entry.set('current_version', self.installed[entry.get('name')])
+                return False
+        entry.set('current_exists', 'false')
+        return False
+
+    def Remove(self, packages):
+        """Remove extra packages."""
+	pkgline = ""
+	for pkg in packages:
+		pkgline += " " + pkg.get('name')
```
Comment (stpierre) [inline src/lib/Bcfg2/Client/Tools/HomeBrew.py]: ``pkgline = " ".join(p.get('name') for p in packages)``

Comment (stpierre) [inline src/lib/Bcfg2/Client/Tools/HomeBrew.py]: Actually, see below.

File: src/lib/Bcfg2/Client/Tools/HomeBrew.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+        self.logger.info("Removing packages: %s" % pkgline)
+        self.cmd.run("/usr/local/bin/brew uninstall %s" % pkgline)
```
Comment (stpierre) [inline src/lib/Bcfg2/Client/Tools/HomeBrew.py]: Use the list version of ``run()`` instead:

    pkg_names = [p.get('name') for p in packages]
    self.logger.info("Removing packages: %s" % pkg_names)
    self.cmd.run(["/usr/local/bin/brew", "uninstall"] + pkg_names)

File: src/lib/Bcfg2/Client/Tools/HomeBrew.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+	"""Install packages."""
+	pkgline = ""
+	for pkg in packages:
+		pkgline += " " + pkg.get('name')
+
+	self.logger.info("About to install packages : " + pkgline)
+
+	try:
+		self.logger.info("Running : brew install %s" % pkgline)
+		self.cmd.run("/usr/local/bin/brew install %s" % pkgline)
```
Comment (stpierre) [inline src/lib/Bcfg2/Client/Tools/HomeBrew.py]: Use a list comprehension to build your list of package names, and the list version of ``run()`` here as well.

File: src/lib/Bcfg2/Client/Tools/HomeBrew.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        self.RefreshPackages()
+        self.extra = self.FindExtra()
+
+    def InstallPackage(self, packages, states):
+	"""Install packages."""
+	pkgline = ""
+	for pkg in packages:
+		pkgline += " " + pkg.get('name')
+
+	self.logger.info("About to install packages : " + pkgline)
```
Comment (stpierre) [inline src/lib/Bcfg2/Client/Tools/HomeBrew.py]: You've already got a much better log message on line 78 below; this can be removed.

**Comment by stpierre** [comment]
Aside from some fairly minor code issues, LGTM.

**Comment by jcftang** [comment]
The comments sound reasonable, I'll fix up the patch and resubmit it.

File: src/lib/Bcfg2/Client/Tools/HomeBrew.py (hunk 4 above)
Comment (jcftang) [inline src/lib/Bcfg2/Client/Tools/HomeBrew.py]: I'm not entirely sure what you suggest to do to resolve this particular issue. Shall I just remove this method?

**Comment by jcftang** [comment]
I think I've addressed most of the comments from the above. However after rebasing to the master branch...

```
(bcfg2)jtang@awesome-o:bcfg2$ bcfg2 -C etc/bcfg2.conf  -v -q  -c tmp
Starting Bcfg2 client run at 1424109059.87
Failed to instantiate tool <class 'Bcfg2.Client.Tools.HomeBrew.HomeBrew'>
Traceback (most recent call last):
  File "/Users/jtang/develop/bcfg2/src/lib/Bcfg2/Client/__init__.py", line 458, in load_tools
    self.tools.append(tool(self.config))
TypeError: __init__() takes exactly 4 arguments (2 given)
Loaded tool drivers:
 Action   Bundle   POSIX    VCS      launchd
The following entries are not handled by any tool:
Package:homebrew:cowsay
```

File: src/lib/Bcfg2/Client/Tools/HomeBrew.py (hunk 4 above)
Comment (stpierre) [inline src/lib/Bcfg2/Client/Tools/HomeBrew.py]: No, just this ``if`` block.

File: src/lib/Bcfg2/Client/Tools/HomeBrew.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+class HomeBrew(Bcfg2.Client.Tools.PkgTool):
+    """Homebrew package support."""
+    name = 'HomeBrew'
+    __execs__ = ["/usr/local/bin/brew"]
+    __handles__ = [('Package', 'homebrew')]
+    __req__ = {'Package': ['name', 'version']}
+    pkgtype = 'homebrew'
+    pkgtool = ('/usr/local/bin/brew install %s', ('%s', ['name']))
+
+    def __init__(self, logger, setup, config):
```
Comment (stpierre) [inline src/lib/Bcfg2/Client/Tools/HomeBrew.py]: Looks like you may have based this file on the ``MacPorts`` driver from 1.3.x, not 1.4.x; lots of function signatures have changed. In particular, the ``__init__`` for client tools only takes a single argument now.

File: src/lib/Bcfg2/Client/Tools/HomeBrew.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+                entry.set('current_version', self.installed[entry.get('name')])
+                return False
+        entry.set('current_exists', 'false')
+        return False
+
+    def Remove(self, packages):
+        """Remove extra packages."""
+        pkg_names = " ".join(p.get('name') for p in packages)
+        self.logger.info("Removing packages: %s" % pkg_names)
+        self.cmd.run(["/usr/local/bin/brew", "uninstall"] + pkg_names)
```
Comment (stpierre) [inline src/lib/Bcfg2/Client/Tools/HomeBrew.py]: ``pkg_names`` shouldn't be a string; this addition will fail, and also it'll take the entire list of names as a single argument, which is not what you want. It should be a list, e.g.:

``pkg_names = [p.get('name') for p in packages]``

File: src/lib/Bcfg2/Client/Tools/HomeBrew.py [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+        self.RefreshPackages()
+        self.extra = self.FindExtra()
+
+    def InstallPackage(self, packages, states):
```
Comment (stpierre) [inline src/lib/Bcfg2/Client/Tools/HomeBrew.py]: ``Install*`` no longer takes a second argument in 1.4.x.

Comment (stpierre) [inline src/lib/Bcfg2/Client/Tools/HomeBrew.py]: Actually, ``PkgTool`` doesn't even use ``InstallPackage``; it automatically generates the install command from ``cls.pkgtool``. This entire method can be removed.

File: src/lib/Bcfg2/Client/Tools/HomeBrew.py (hunk 9 above)
Comment (jcftang) [inline src/lib/Bcfg2/Client/Tools/HomeBrew.py]: Yes, that seems to be the case.

**Comment by jcftang** [comment]
The proposed changes have been made

**Comment by jcftang** [comment]
Oops, I need to cleanup the last commit

**Comment by solj** [comment]
Can you also add this to the 1.4.0pre2 release notes?

**Comment by solj** [comment]
:+1:

File: src/lib/Bcfg2/Client/Tools/HomeBrew.py [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+                                 "list", "--versions"]).stdout.splitlines()
+        self.installed = {}
+        for pkg in pkgcache:
+            pkgname, version = pkg.strip().split()
+            self.logger.debug(" pkgname: %s version: %s" % (pkgname, version))
+            self.installed[pkgname] = version
+
+    def VerifyPackage(self, entry, _):
+        """Verify Package status for entry."""
+        self.logger.info("VerifyPackage: %s : %s" % (entry.get('name'),
```
Comment (stpierre) [inline src/lib/Bcfg2/Client/Tools/HomeBrew.py]: Log this at debug.

File: src/lib/Bcfg2/Client/Tools/HomeBrew.py [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+                                                     entry.get('version')))
+        if 'version' not in entry.attrib:
+            self.logger.info("Cannot verify unversioned package %s" %
+                             entry.attrib['name'])
+            return False
```
Comment (stpierre) [inline src/lib/Bcfg2/Client/Tools/HomeBrew.py]: This if block is unreachable.

**Comment by stpierre** [comment]
Github seems to have eaten a couple of my older comments -- maybe it got confused in the rebase.

**Pull request merged**